lib/logthrdestdrv.c

```diff
@@ -325,16 +363,40 @@ _message_became_available_callback(gpointer user_data)
     iv_event_post(&self->wake_up_event);
 }
 
+static void
+_schedule_restart_on_suspend_timeout(LogThreadedDestDriver *self)
+{
+  iv_validate_now();
+  self->timer_reopen.expires  = iv_now;
+  self->timer_reopen.expires.tv_sec += self->time_reopen;
+  iv_timer_register(&self->timer_reopen);
+}
+
+static void
+_schedule_restart_on_flush_timeout(LogThreadedDestDriver *self)
+{
+  iv_validate_now();
+  self->timer_flush.expires = self->last_flush_time;
+  timespec_add_msec(&self->timer_flush.expires, self->flush_timeout);
+  iv_timer_register(&self->timer_flush);
+}
+
 static void
 _schedule_restart(LogThreadedDestDriver *self)
 {
   if (self->suspended)
-    {
-      iv_validate_now();
-      self->timer_reopen.expires  = iv_now;
-      self->timer_reopen.expires.tv_sec += self->time_reopen;
-      iv_timer_register(&self->timer_reopen);
-    }
+    _schedule_restart_on_suspend_timeout(self);
+  else
+    iv_task_register(&self->do_work);
+}
+
+static void
+_schedule_restart_on_next_flush(LogThreadedDestDriver *self)
+{
+  if (self->suspended)
+    _schedule_restart_on_suspend_timeout(self);
+  else if (!_flush_time_has_elapsed(self))
+    _schedule_restart_on_flush_timeout(self);
```

Comment:
Here in the `!_flush_time_has_elapsed(self)` branch we should still call `iv_task_register(&self->do_work);`, if there are messages in the queue, we should handle them.

Reply:
@furiel: flush_handler and work_handler is the same : `_perform_work`, so I think it is correct, but maybe I'm wrong.

Reply:
They almost do the same, but `flush_handler` executes `_perform_work` only `flush_timeout` later. But when we are in the `!suspended` branch, we should have a message to be handled. It is possible that is the last message missing from a batch to flush, which would delay the batch flush unnecessarily.

Reply:
do_work would wake up immediately and not when the message arrives. This means that we would spin in the thread, as we would wake up immediately, finding that the queue is empty.

The way this is handled is that log_queue_check_items() registers a callback function whenever there is NO message in the queue and that gets called if one is added, this function is _message_became_available_callback() and that wakes up our thread immediately.

So in summary:
1) this function is only called when we have some items in the batch but no items in the queue
2) in this case message_became_available_callback() is registered and would be called by the LogQueue
3) any new message thus wakes up the thread, which would pull all of the messages it finds there.
4) if the batch is still not complete we would repeat the same process until either the flush timeout elapses, or we get enough items  to fill a batch.

I think this should be what we expect. I agree though that the wakeup logic is a bit messy here. I tried a few alternatives, but none seemed to read better :(